libraries/lib-xml/XMLAttributeValueView.cpp

```diff
@@ -0,0 +1,415 @@
+   std::string_view defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+std::string XMLAttributeValueView::ToString() const
+{
+   switch (mType)
+   {
+   case XMLAttributeValueView::Type::Null:
+      return {};
+   case XMLAttributeValueView::Type::SignedInteger:
+      return std::to_string(mInteger);
+   case XMLAttributeValueView::Type::UnsignedInteger:
+      return std::to_string(static_cast<uint64_t>(mInteger));
+   case XMLAttributeValueView::Type::Float:
+      return std::to_string(mFloat);
+   case XMLAttributeValueView::Type::Double:
+      return std::to_string(mDouble);
+   case XMLAttributeValueView::Type::StringView:
+      return std::string(mStringView.Data, mStringView.Length);
+   }
+
+   return {};
+}
+
+wxString XMLAttributeValueView::ToWString() const
+{
+   switch (mType)
+   {
+   case XMLAttributeValueView::Type::Null:
+      return {};
+   case XMLAttributeValueView::Type::SignedInteger:
+      return wxString() << mInteger;
+   case XMLAttributeValueView::Type::UnsignedInteger:
+      return wxString() << mInteger; static_cast<uint64_t>(mInteger);
+   case XMLAttributeValueView::Type::Float:
+      return wxString() << mFloat;
+   case XMLAttributeValueView::Type::Double:
+      return wxString() << mDouble;
+   case XMLAttributeValueView::Type::StringView:
+      return wxString::FromUTF8(mStringView.Data, mStringView.Length);
+   }
+
+   return {};
+}
+
+template <typename ResultType>
+bool CheckInteger(ResultType& output, int64_t value) noexcept
+{
+   constexpr int64_t minValue = std::numeric_limits<ResultType>::min();
+   constexpr int64_t maxValue = std::numeric_limits<ResultType>::max();
+
+   if (minValue <= value && value <= maxValue)
+   {
+      output = static_cast<ResultType>(value);
+      return true;
+   }
+
+   return false;
+}
+
+template <typename ResultType>
+bool CheckInteger(ResultType& output, uint64_t value) noexcept
+{
+   constexpr uint64_t maxValue = std::numeric_limits<ResultType>::max();
+   const uint64_t unsignedValue = static_cast<uint64_t>(value);
+
+   if (unsignedValue <= maxValue)
+   {
+      output = static_cast<ResultType>(unsignedValue);
+      return true;
+   }
+
+   return false;
+}
+
+template <typename ResultType>
+bool XMLAttributeValueView::GetInteger(ResultType& value) const noexcept
+{
+   static_assert(std::is_integral_v<ResultType>);
+
+   if (mType == Type::SignedInteger)
+   {
+      return CheckInteger(value, mInteger);
+   }
+   else if (mType == Type::UnsignedInteger)
+   {
+      return CheckInteger(value, static_cast<uint64_t>(mInteger));
+   }
+   else if (mType == Type::StringView)
+   {
+      const char* end = mStringView.Data + mStringView.Length;
+
+      ResultType tempValue = {};
+
+      const auto result = FromChars(mStringView.Data, end, tempValue);
+
+      if (result.ec == std::errc() && result.ptr == end)
+      {
+         value = tempValue;
+         return true;
+      }
+   }
+
+   return false;
+}
```

Comment:
This entire source file makes me want to find a nicer rewrite with variant and Visit and less repetition...  but not now

Reply:
There will be a huge set of nested ifs by a type, won't it? So the amount of code will be roughly the same, the compilation speed will be much worse though.

Get can be rewritten easily using a template, but again, the cost will be the compilation time again.

Reply:
(You know, like)

```
template<typename T>
bool Get(T& value) const
{
  if constexpr(std::is_integral_v<T>) 
    return GetInteger(value);
 else ...
}
```

Reply:
I like avoidance of repetition where I can get it with templates, with its advantages for maintainability.  Maybe I should care more about build time.  But you know... “Lisp programmers know the value of everything and the cost of nothing.”  Substitute “meta” for “Lisp”

Reply:
Templates have a significant cost attached. They are slow to compile; they bloat the executables and libraries. In this case, having all the internal of the class exposed seems a bit much to me. There are places where they are extremely helpful, but I really don't think that this is the one.